### DISCUSSION TITLE
The right way to use Prisma with Cloudflare Workers

### DISCUSSION THREAD

**Discussion by sakettawde** [prisma/prisma#23762] [Q&A] [answered]
### Question

So over the weekend wanted to run some tests with Prisma running in Cloudflare workers (an edge serverless environment) with different databases. In the process wanted to set it up like a real world application. I ran into the following issue.

Since Cloudflare Workers send their environment variables via a context object in the function call, I cannot initialise the prisma client outside the handler. While this didn't seem to be an apparent issue on the surface while testing with multiple simultaneous requests, looking into the logs, I saw this:
`warn(prisma-client) This is the 10th instance of Prisma Client being started. Make sure this is intentional.`

I looked around and realised no one had an example to show a real world setup for this tech stack. Then I came across this article: https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-dev-practices

Inspired by it, I put together something simple. Sharing an oversimplification of what is in my code:

```
const global = {}

export const prismaClient = (c) => {
  if (global.prisma) return global.prisma

  const libsql = createClient({
    url: c.env.LIBSQL_DB_URL,
    authToken: c.env.LIBSQL_DB_AUTH_TOKEN,
  })

  const adapter = new PrismaLibSQL(libsql)
  const prisma = new PrismaClient({ adapter })

  global.prisma = prisma
  return prisma
}
```

This triggered another error from Cloudflare Workers in production:
`Error: Cannot perform I/O on behalf of a different request. I/O objects (such as streams, request/response bodies, and others) created in the context of one request handler cannot be accessed from a different request's handler. This is a limitation of Cloudflare Workers which allows us to improve overall performance.`

What is the correct way to proceed here?

**Comment by sakettawde** [comment, marked as answer]
@nikolasburk reached out on a thread where I had shared the same on Twitter.
Here is the link: https://x.com/nikolasburk/status/1777255194964771245

Closing this for now

**Comment by matteo-hertel** [comment]
I came across this issue looking for the same answer. Thanks for starting this conversation!

There was tons of context and links so let me summarise this for the next person coming across this issue:

- as of right now the right way of instantiating the prisma client in CF workers is INSIDE the handle function. From the performance POV, the added latency is minimal.

- doing so you might get a warning `warn(prisma-client) This is the 10th instance of Prisma Client being started.` you can safely ignore this, it will be removed soon (or removed by the time you read this)

- if you're using an adapter (for PG or SQLite) you don't need to user the client from the `/edge` import. the docs will be updated soon

The question that wasn't answered is how to handle the creation of the client and adding that to the the `globalThis` scope.

I tried that and in prod I get the `Error: Cannot perform I/O on behalf of a different request. I/O objects (such as streams, request/response bodies, and others) created in the context of one request handler cannot be accessed from a different request's handler. This is a limitation of Cloudflare Workers which allows us to improve overall performance.` error.

Cloudflare  mentions this in their docs. Do not add anything to the global scope as a worker can handle multiple requests.

@sakettawde have you find a solution for the global scope issue?

it's all fine when the worker is a hello world. When you have a production application that might have a layered architecture is not so easy.

For example (pseudocode):

```
handler
|
-> FriendsService
               |
               -> UserService
                            |
                            -> queries -> in this file I need the prisma client to get a list of users
```
I add the client to the global scope I can access it in my service but I get random I/O errors

Is there a better way to do it then pass the DB instance around?

**Reply by whilemouse** [reply]
Did you solve this problem? I am having the same problem.

**Reply by matteo-hertel** [reply]
Kind of.
I’ve re-architected my layers to pass a database instance all the way down to where is needed.
Not ideal in my view, but I used opportunity to create a businessLogicContext object and pass that around.
That made my code more testable and worked around this problem.
I’ve also added cloudflare hyperdrive.
The end result is: my db instance is created in the main handler and passed around my code to use it whenever I need it.

I never had that issue before but it took a whole bunch of work. 
hope it helped

**Comment by sunnylqm** [comment]
so you can use global.prisma but need to refresh it everytime in a new request

**Reply by polaroidkidd** [reply]
would I still not have to create it in the handle function and then pass it around as @matteo-hertel  does?

**Reply by sunnylqm** [reply]
something like https://github.com/nick-cjyx9/Random-Elysia/blob/main/backend/src/index.ts#L12

**Reply by sunnylqm** [reply]
```js
export default {
  async fetch(request: Request, env: Env) {
    global.prisma = new PrismaClient({ adapter });
  },
}
```

**Reply by ronakmaheshwari** [reply]
need some help here
